kbp/config/config_v1.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package config
+
+import (
+	"crypto/sha512"
+	"encoding/hex"
+	"sync"
+)
+
+// MakeSaltedSHA512PasswordHex takes concatenated salt and password, calculate
+// SHA512 of it, and return the result encoded in hex.
+func MakeSaltedSHA512PasswordHex(password string, salt string) string {
+	s := sha512.Sum512([]byte(salt + password))
+	return hex.EncodeToString(s[:])
+}
+
+// UserV1 defines a user's hashed password for the V1 config.
+type UserV1 struct {
+	PasswordSHA512Hex string `json:"password_sha512_hex"`
+	PasswordSalt      string `json:"password_salt"`
+}
+
+// Authenticate returns true if password matches u, or false if not.
+func (u *UserV1) Authenticate(password string) bool {
+	return MakeSaltedSHA512PasswordHex(
+		password, u.PasswordSalt) == u.PasswordSHA512Hex
+}
+
+// AccessControlV1 defines an access control list (ACL) for the V1 config.
+type AccessControlV1 struct {
+	// WhitelistAdditionalPermissions is a map of username -> permissions that
+	// defines a list of additional permissions that authenticated users have
+	// in addition to AnonymousPermissions
+	WhitelistAdditionalPermissions map[string]string `json:"whitelist_additional_permissions"`
+	// AnonymousPermissions is the permissions for
+	// unauthenticated/anonymous requests.
+	AnonymousPermissions string `json:"anonymous_permissions"`
+}
+
+// V1 defines a V1 config. Public fields are accessible by `json`
+// encoders and decoder. On first call to GetPermission* methods, it
+// initializes an internal ACL checker. Any changes to the ACL fields
+// afterwards have no effect.
+type V1 struct {
+	Common
+
+	// Users is a username -> UserV1 map that defines how users should be
+	// authenticated.
+	Users map[string]UserV1 `json:"users"`
+
+	// DefaultACL defines the ACL that should be used for paths that no ACL has
+	// been configured for.
+	DefaultACL AccessControlV1 `json:"default_acl"`
+	// ACLs is a path -> AccessControlV1 map that defines ACLs for different
+	// paths.
+	ACLs map[string]AccessControlV1 `json:"acls"`
+
+	initOnce          sync.Once
+	aclChecker        *aclCheckerV1
+	aclCheckerInitErr error
+}
+
+// DefaultV1 returns a default V1 config, which allows anonymous read to
+// everything.
+func DefaultV1() *V1 {
+	return &V1{
+		Common: Common{
+			Version: Version1Str,
+		},
+		DefaultACL: AccessControlV1{
+			AnonymousPermissions: "read",
```

Comment:
Why isn't "read,list" the right anonymous permissions?  We don't want to support the `keybase.pub/chris`-style navigation by default?  (If not, why do we even have a "list" permision?)

Reply:
> by default

I'm OK with turning that on by default.

> If not, why do we even have a "list" permision?

Hmm I'm not sure I understand the logic here. Why would we not have a "list" permission if we don't have that on by default? User might want to turn on listing only for a particular subdir right?

Reply:
It seems like if we were to implement directory listings, we could replace `keybase.pub/chris` with a kbpbot.  But we can't do that if the default permissions do not include listing, because it would require each user to add a config file to their public directory to get the same permissions we have today.

If we are not going to implement directory listings, then there's no need for a "list" permission at all.

So I don't understand what middle ground you're targeting with these defaults.

Reply:
Oh I see; that makes sense. I missed the point where public TLF backed sites don't have config files. The middle ground I'm targeting is `strib.io/photos/secret` where `photos` can be listed, but `secret` cannot so you can have `strib.io/photos/secret/blahblah-secret-url/`.

Go's `FileServer` implementation already supports listing, so we don't really need to implement anything for it, assuming we are happy with what Go provides. If we are ever gonna implement keybase.pub/chris styled listing, it would be somewhat different I think, since it'd require integrating with other parts of the API server like login, etc.

Reply:
> The middle ground I'm targeting is strib.io/photos/secret where photos can be listed, but secret cannot so you can have strib.io/photos/secret/blahblah-secret-url/.

But why is it the right idea to require the user to override the defaults for `photos`, but not for `photos/secret`?  Wouldn't they already have to set an explicit permissions in `photos/secret`, in order to override the `photos` permissions (which, in your case, would have to have been added in order to allow for listing)?

> If we are ever gonna implement keybase.pub/chris styled listing, it would be somewhat different I think, since it'd require integrating with other parts of the API server like login, etc.

I don't follow, what about keybase.pub listing requires login?  I thought it was just a formatting difference.

Reply:
> Wouldn't they already have to set an explicit permissions in photos/secret, in order to override the photos permissions

Not as in this implementation, where `photos/secret` permissions is completely separate from `photos`, since unspecified paths just use the default ACL. But I'm realizing that's a bad idea, since user might just accidentally put in directories in `photos` and they would get automatic defaultACL permission even if `photos` has a more restrictive one.

I think the right idea is to inherit parent permission instead of a "default" one. The `Config.DefaultACL` can then go away. And if "/" doesn't appear in the ACLs field, we'd construct using `DefaultV1` for "/". How does that sound?

> I don't follow, what about keybase.pub listing requires login? I thought it was just a formatting difference.

Ah you're right! Somehow I thought keybase.pub populates usernames like on keybase.io. But it doesn't, so I agree it'd just be formatting difference now.

Reply:
> I think the right idea is to inherit parent permission instead of a "default" one. The Config.DefaultACL can then go away.

Yes, that sounds right.  That's how I read the code but I messed up.  A directories permissions should definitely apply by default to all its subdirectories as well.

Reply:
Oh actually you didn't read it wrong. If user defines permission for `/` only, then anything under that automatically inherits it. But it's an issue if user only defines `/` and `/a/b`, but not `/a`, in which case `/a` would get the default ACL rather than the one from `/`. It's more like a bug here :)

Reply:
I'm not sure I agree that a parent's permissions could be automatically inherited by children. IMO the default permissions should be "nothing", and you have to explicitly express paths that are allowed to be read and/or listed.

Then again, `.htaccess` has `Allow,Deny`, and `Deny,Allow` to allow you to dictate those semantics explicitly. The question remains whether we want our default to be more or less permissive.

Reply:
> I'm not sure I agree that a parent's permissions could be automatically inherited by children.

More precisely, it's the parent's ACL that's inherited by children. In other words, if `/a/b` has an ACL that disallows anything from anonymous, `/a/b/c` automatically gets that too, even if `/a` has a more permissive ACL. This is so that if you copy a site consisting of a bunch of directories and sub directories into a place, you don't have to define an ACL for each of the subdirectories.